### DISCUSSION TITLE
How to conditionally Consume Data?

### DISCUSSION THREAD

**Discussion by NexRX** [rwf2/Rocket#2526] [Questions] [answered]
So, in Rocket 0.5-rc3, I'm trying to create my own type that Implements FromData so I can implement something like TUS protocol (Basically, resumable uploads via chunking the data) among other such things.

However, If the body of the request isn't used in someway via the Data type. I've found the HTTP connection is closed/aborted about more than 50% of the time.

So doing something like this with rocket: 
```rust
#[rocket::async_trait]
impl<'r> FromData<'r> for FileFull<'r> {
    type Error = Message;

    async fn from_data(
        req: &'r Request<'_>,
        mut data: rocket::data::Data<'r>,
    ) -> DataOutcome<'r, Self> {
        let content_length = req.guard::<ContentLength>().await.unwrap();

        if content_length.0 > 30.megabytes() {
            return DataOutcome::Failure((
                Status::BadRequest,
                Message::bad_request(Some(vec![
                    "Content-Length is too long. Cannot be longer than 30MB".to_string(),
                ])),
            ));
        }

        ...
    }
}
```
Results in the  connection _sometimes_ aborted. Weirdly enough, if the Data/Body is small enough (less than 1MB) and you don't use it, the connection doesn't get closed.

But if I remove the (example) condition and use the data contents in it's entirety (regardless of size), it's completely, fine.

I have done so much looking into previous issues, I thought maybe I was doing some blocking operations and the body streaming was failing because of that, but I am not, as you can see in the example.

Is anyone familiar with why this is happening and how I can avoid these connection drop?
Many thanks!

**Comment by SergioBenitez** [comment, marked as answer]
You're observing a security mechanism in Rocket that aims to prevent certain kinds of (D)DoS attacks. In particular, ones where an attacker seeks to tie up all network bandwidth through multiple connections that send ~infinite amounts of data.

To prevent this kind of attack, an application server needs to limit the amount of data it's willing to accept from a client. Rocket enforces this statically via `Data::open()`, preventing the class of attack entirely. But this necessitates a decision on what to do in the case that a client _does_ send more data than expected. Since messages are serialized on a single connection in HTTP 1.1, we _need_ to read a complete message (and all of its data) in order to read the next message, so we can't simply stop a client from sending us data _and_ move on to the next message: we either read the entire message or abort the connection. So Rocket does a bit of the former and then the latter. It tries to salvage the connection by reading and dumping _some_ unexpected data, but if there's _still_ data remaining, then we have to close the connection.

I hope this explain what's happening. If this isn't documented anywhere, we should do so.

**Reply by NexRX** [reply]
Aboslutely understand why now. I knew this was the expected behaviour with Limits but I didn't see any documentation about in-general handling of Data like this.

Is there an elegant way to reject the request in these situations?

Because currently im doing access token (JWT) authentication with short lived tokens and now I have this problem.
When one of these tokens expires and I make an attempt to upload a file, the upload will fail but because the connection can just be aborted, the user has no contextual response saying why it failed (like a 401 status). 

There are client work-around but I was wondering if there is someway possible to with Http2 or so on to let the user know why it failed.

**Reply by SergioBenitez** [reply]
> Aboslutely understand why now. I knew this was the expected behaviour with Limits but I didn't see any documentation about in-general handling of Data like this.

I'd happily accept a PR to ameliorate this situation!

> Is there an elegant way to reject the request in these situations?

I don't believe so. I don't believe browsers will listen for early responses, even when the spec suggests that they do. 

 
> Because currently im doing access token (JWT) authentication with short lived tokens and now I have this problem.
> 
> When one of these tokens expires and I make an attempt to upload a file, the upload will fail but because the connection can just be aborted, the user has no contextual response saying why it failed (like a 401 status). 
> 
> There are client work-around but I was wondering if there is someway possible to with Http2 or so on to let the user know why it failed.

The easiest and most reliable approach is to use a pre-flight auth check request, say a simple HEAD. Ask the server if such a request _would_ work, and then send the request. Alternatively just handle the aborted connection gracefully at the client. 

In theory one could handle this gracefully with HTTP/2, but that would still require browsers to interpret out-of-order messages on a single stream, which they may or may not do. HTTP/2 is largely an improvement at the physical protocol level, not the logical level, which is where this problem arises. But more importantly, such a solution would only work for client operating over HTTP/2, which seems like a major drawback, at least for now.

**Reply by NexRX** [reply]
I kept it simple and implemented a head request to verify if a request should work like how you said. Thanks for the explination! Good to know Rocket was also saving my arse in reality.
